### PR TITLE
New Component: List Input

### PR DESCRIPTION
**Pull request by domoscargin**
https://trello.com/c/P13uxriu/135-5-add-list-input-component-to-digital-marketplace-govuk-frontend

This builds on the work done in #35

I've moved some of the app-focused changes to a separate PR, which this PR now requires: #148

### CODE REVIEW

File: src/digitalmarketplace/common.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/**
+ * TODO: Ideally this would be a NodeList.prototype.forEach polyfill
+ * This seems to fail in IE8, requires more investigation.
+ * See: https://github.com/imagitama/nodelist-foreach-polyfill
+ */
```
Comment (gidsg) [inline src/digitalmarketplace/common.js:5]: do we still need to support IE8? Have you tried the polyfills at https://developer.mozilla.org/en-US/docs/Web/API/NodeList/forEach#Polyfill ?

Comment (gidsg) [reply src/digitalmarketplace/common.js:5]: Alteratively we could do this without polyfill using one one the hacks described in https://css-tricks.com/snippets/javascript/loop-queryselectorall-matches/ ?

File: src/digitalmarketplace/components/list-input/list-input.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        $input.setAttribute('disabled', 'true')
+      }
+      numberOfVisibleEmptyItems += 1
+    } else {
+      $item.classList.remove('dm-list-input__item--hidden')
+      $removeButton.classList.remove('dm-list-input__item-remove--hidden')
+      numberOfFilledInItems += 1
+    }
+  })
+  this.$allVisibleItems = this.$module.querySelectorAll('.dm-list-input__item:not(.dm-list-input__item--hidden)')
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.js:72]: seems like this `:not` selector is not supported in IE? https://caniuse.com/#search=%3Anot Do we need to find an alternative way?

File: src/digitalmarketplace/components/list-input/list-input.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        $errorContainer.classList.remove('dm-list-input__item--error')
+        $errorContainer.querySelector('.govuk-error-message').remove()
+        var $errorInput = $errorContainer.querySelector('.govuk-input--error')
+        var inputId = $errorInput.getAttribute('id')
+        var inputDescribedBy = $errorInput.getAttribute('aria-describedby')
+        $errorInput.setAttribute('aria-describedby', inputDescribedBy.replace(inputId + '-error', ''))
+        $errorContainer.querySelector('.govuk-input--error').classList.remove('govuk-input--error')
+      }
+
+      // Hide Remove buttons if there are only one item left
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.js:105]:
```suggestion
      // Hide Remove buttons if there is only one item left
```

File: src/digitalmarketplace/common.js (hunk 1 above)
Comment (domoscargin) [reply src/digitalmarketplace/common.js:5]: This file was previously copied from https://github.com/alphagov/govuk-frontend/blob/master/src/govuk/common.js, which does have some support for IE8. I figured I'd keep it in place for maximum compatibility, but I guess since that method is 2+ years old, and we don't have explicit support for IE8 on DMp, I'll swap this out for a polyfill.

File: src/digitalmarketplace/components/list-input/list-input.js (hunk 2 above)
Comment (domoscargin) [reply src/digitalmarketplace/components/list-input/list-input.js:72]: Oof, good catch

File: src/digitalmarketplace/components/list-input/list-input.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+  }
+  while (sibling) {
+    if (sibling.matches(selector)) return sibling
+    sibling = (direction === 'next') ? sibling.nextElementSibling : sibling = sibling.previousElementSibling
+  }
+}
+
+function ListInput ($module) {
+  this.$module = $module
+  this.$allVisibleItems = $module.querySelectorAll('.dm-list-input__item:not(.dm-list-input__item--hidden)')
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.js:24]: is this use of `:not` safe?

File: src/digitalmarketplace/components/list-input/list-input.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+  this.visibleItemClass = 'dm-list-input__item--visible'
+  this.itemInputClass = 'dm-list-input__item-input'
+  this.itemErrorClass = 'dm-list-input__item--error'
+  this.itemErrorMsgClass = 'govuk-error-message'
+  this.itemCounterClass = 'dm-list-input__counter'
+  this.inputErrorClass = 'govuk-input--error'
+  this.removeButtonClass = 'dm-list-input__item-remove'
+  this.hiddenRemoveButtonClass = 'dm-list-input__item-remove--hidden'
+
+  this.formgroupErrorClass = 'govuk-form-group--error'
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.js:40]: Can we make this `formGroupErrorClass` for consistency with the other variables?

File: src/digitalmarketplace/components/list-input/list-input.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+  // Check for module
+  if (!this.$module) {
+    return
+  }
+
+  this.hideEmptyItems()
+  this.updateAllCounters()
+
+  this.bindRemoveClickEvent()
+
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.js:55]: linebreaks are inconsistent here, maybe remove these?

File: src/digitalmarketplace/components/list-input/list-input.test.js [hunk 7]
```diff
@@ -0,0 +1,280 @@
+const configPaths = require('../../../../config/paths.json')
+const PORT = configPaths.ports.test
+
+// URLS
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.test.js:4]:
```suggestion
// URLs
```

File: src/digitalmarketplace/components/list-input/list-input.js (hunk 4 above)
Comment (domoscargin) [reply src/digitalmarketplace/components/list-input/list-input.js:24]: 🤦  I promise I'm not missing these on purpose.

File: src/digitalmarketplace/components/list-input/list-input.yaml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  type: object
+  required: true
+  description: Options for the fieldset component (e.g. legend).
+- name: hint
+  type: object
+  description: Options for the hint component.
+- name: id
+  type: string
+  required: true
+  description: Used as an `id` in the HTML for the list-input as a whole, and also as a prefix for the `id`'s of the hints & input fields'
```
Comment (gidsg) [inline src/digitalmarketplace/components/list-input/list-input.yaml:16]: stray apostrophe. 
```suggestion
  description: Used as an `id` in the HTML for the list-input as a whole, and also as a prefix for the `id`s of the hints & input fields'
```

File: src/digitalmarketplace/components/list-input/template.njk [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      html: params.errorMessage.html,
+      text: params.errorMessage.text,
+      visuallyHiddenText: params.errorMessage.visuallyHiddenText
+    }) | indent(2) | trim }}
+  {% endif %}
+  {% call govukFieldset({
+    describedBy: params.fieldset.describedBy if params.fieldset.describedBy else describedBy,
+    classes: params.fieldset.classes,
+    legend: {
+      html: params.fieldset.legend.text + ('<span class="dm-list-input__optional__optional"> (Optional)</span>' if params.optional) ,
```
Comment (lfdebrux) [inline src/digitalmarketplace/components/list-input/template.njk:38]:
```suggestion
      html: params.fieldset.legend.text + ('<span class="dm-list-input__optional"> (Optional)</span>' if params.optional) ,
```

Comment (lfdebrux) [reply src/digitalmarketplace/components/list-input/template.njk:38]: I'm not sure whether having the optional text here makes the most sense; it might make more sense to define it closer to the point of use i.e. in the function that transforms the content loader question? What do you think?

File: src/digitalmarketplace/components/list-input/template.njk [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        {% if item.errorMessage %}
+        {{ govukErrorMessage({
+          id: errorId,
+          classes: 'dm-list-input-error-message  ' + item.errorMessage.classes if item.errorMessage.classes,
+          attributes: item.errorMessage.attributes,
+          html: item.errorMessage.html,
+          text: item.errorMessage.text,
+          visuallyHiddenText: item.errorMessage.visuallyHiddenText
+        }) | indent(2) | trim }}
+        {% endif %}
```
Comment (lfdebrux) [inline src/digitalmarketplace/components/list-input/template.njk:79]: In the GOV.UK Design System error messages are after the question text and hint text [1](https://design-system.service.gov.uk/components/error-message/#how-it-works), is there a reason not to follow that design rule for this component?

**Comment by lfdebrux** [comment]
Playing around with this in the review app it looks good, modulo some questions I have above.

This branch has quite a lot of commits though; would you be able to split out some of the non-list input related changes into a separate PR? Or if you want I can do that.

File: src/digitalmarketplace/components/list-input/template.njk (hunk 9 above)
Comment (domoscargin) [reply src/digitalmarketplace/components/list-input/template.njk:38]: I think I agree, though it may depend on what approach we take to https://trello.com/c/YXsBOrDK/110-3-replace-text-boxes-with-govuk-frontend-text-input-component-in-create-a-dos-opportunity-journey

If we end up dropping questions as-is into templates on this repo, then this approach might work. If we transform things in content-loader or utils, then this isn't optimal.

File: src/digitalmarketplace/components/list-input/template.njk [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+      html: params.errorMessage.html,
+      text: params.errorMessage.text,
+      visuallyHiddenText: params.errorMessage.visuallyHiddenText
+    }) | indent(2) | trim }}
+  {% endif %}
+
+  <div class="dm-list-input__item-container" role="region" aria-live="assertive">
+    {% for i in range(0, params.maxItems) -%}
+      {%- set item = params.items[i] if params.items[i] else {} -%}
+      {%- set entryItemId = params.id + '-' + i -%}
```
Comment (lfdebrux) [inline src/digitalmarketplace/components/list-input/template.njk:60]: We should use `loop.index` here the same way that govuk-frontend does so that the id numbers are 1 indexed as other components (for an example see [checkboxes component](https://github.com/alphagov/govuk-frontend/blob/master/src/govuk/components/checkboxes/template.njk#L61-L66))

File: src/digitalmarketplace/components/list-input/template.njk (hunk 10 above)
Comment (domoscargin) [reply src/digitalmarketplace/components/list-input/template.njk]: There's two places for error messages on this component:

I've moved the "component-level" error message as per that guidance (which I had missed - thank you!). This is for an error that affects the whole component (ie: "Enter at least one requirement").

The error message above is intended to be used at item level - tie an error to an input (ie: "too few characters", etc). Previously the `govukInput` it should attach to was hardcoded, but now that it's not, I'm going to tweak it to use the `errorMessage` property.

**Comment by lfdebrux** [comment]
Is there a way we can test what gets submitted to the server when the browser POSTs the data from one of these components?

I've noticed that it's possible to get the `id` numbers out of order by deleting and adding fields, and while I don't think it makes a difference to the order of the data submitted, it would be good to check...

**Comment by domoscargin** [comment]
> Playing around with this in the review app it looks good, modulo some questions I have above.
> 
> This branch has quite a lot of commits though, meaning reviewing is taking a while; would you be able to split out some of the non-list input related changes into a separate PR? Or if you want I can do that.

I've swapped out some commits to #148 

That now needs to be approved and merged before this branch passes.

**Comment by domoscargin** [comment]
> Is there a way we can test what gets submitted to the server when the browser POSTs the data from one of these components?
> 
> I've noticed that it's possible to get the `id` numbers out of order by deleting and adding fields, and while I don't think it makes a difference to the order of the data submitted, it would be good to check...
> 
> (Edited to add: it doesn't have to be a unit test!)

Hmm, I've just been hard-coding a form wrapper when I want to check and it looks OK. Perhaps I can add some kind of "wrap" property to the yaml and create an example within a form.

**Comment by lfdebrux** [comment]
> Hmm, I've just been hard-coding a form wrapper when I want to check and it looks OK. Perhaps I can add some kind of "wrap" property to the yaml and create an example within a form.

We can create examples without YAML hoops by adding templates to `app/views/examples`, like in [govuk-frontend](https://github.com/alphagov/govuk-frontend/tree/master/app/views/examples) or PR #126

File: src/digitalmarketplace/all.js [hunk 12]
```diff
@@ -1,10 +1,20 @@
...
+import ListInput from './components/list-input/list-input'
 
 function initAll (options) {
+  // Set the options to an empty object by default if no options are passed.
+  options = typeof options !== 'undefined' ? options : {}
+
+  // Allow the user to initialise Digital Marketplace GOV.UK Frontend in only
+  // certain sections of the page
+  // Defaults to the entire document if nothing is set.
+  var scope = typeof options.scope !== 'undefined' ? options.scope : document
```
Comment (lfdebrux) [inline src/digitalmarketplace/all.js:16]: Why is this needed? Asking more out of curiosity

Comment (domoscargin) [reply src/digitalmarketplace/all.js:16]: It looks like it was following the structure of govuk-frontend, so you can plug it in to certain bits of a page rather than the whole thing.

I think it's probably not needed for this particular repo, since we always use it on the whole page, but thought I'd leave it in to allow that functionality.

Can always revert if preferred - though as it stands we're passing in the options arg and not using it.
